analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsLogFailures.cs

```diff
@@ -0,0 +1,189 @@
+            if (catchClause.Block is { } block)
+            {
+                walker.Visit(block);
+            }
+
+            if (!walker.HasValidLoggerCall && catchClause.Filter?.FilterExpression is { } filterExpression)
+            {
+                walker.Visit(filterExpression);
+            }
+
+            return walker.HasValidLoggerCall
+                ? null
+                : walker.InvalidLoggerInvocations;
+        }
+
+        private static IMethodSymbol AzureFunctionEntryPoint(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            return node.Ancestors().OfType<MethodDeclarationSyntax>().FirstOrDefault() is { } method
+                && method.AttributeLists.Any() // FunctionName has [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)] so there must be an attribute
+                && model.GetDeclaredSymbol(method, cancellationToken) is { } methodSymbol
+                && model.Compilation.GetTypeByMetadataName(KnownType.Microsoft_Azure_WebJobs_FunctionNameAttribute.TypeName) is { } functionNameAttribute
+                && methodSymbol.GetAttributes().Any(a => a.AttributeClass.Equals(functionNameAttribute))
+                ? methodSymbol
+                : null;
+        }
+
+        private sealed class LoggerCallWalker : SafeCSharpSyntaxWalker
+        {
+            private static readonly int[] ValidLogLevel = new[]
+            {
+                3, // Warning
+                4, // Error
+                5, // Critical
+            };
+
+            private readonly ImmutableHashSet<ExpressionSyntax>.Builder invalidIvocationsBuilder = ImmutableHashSet.CreateBuilder<ExpressionSyntax>();
+
+            public LoggerCallWalker(ILanguageFacade languageFacade, SemanticModel model, ITypeSymbol iLoggerSymbol, CancellationToken cancellationToken)
+            {
+                Language = languageFacade;
+                Model = model;
+                ILogger = iLoggerSymbol;
+                CancellationToken = cancellationToken;
+                LoggerExtensions = new Lazy<INamedTypeSymbol>(() => Model.Compilation.GetTypeByMetadataName(KnownType.Microsoft_Extensions_Logging_LoggerExtensions.TypeName));
+                ILogger_Log = new Lazy<IMethodSymbol>(() => ILogger.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(x => x.Name == "Log"));
+            }
+
+            private ILanguageFacade Language { get; }
+            private SemanticModel Model { get; }
+            private ITypeSymbol ILogger { get; }
+            private CancellationToken CancellationToken { get; }
+            private Lazy<INamedTypeSymbol> LoggerExtensions { get; }
+            private Lazy<IMethodSymbol> ILogger_Log { get; }
+            public bool HasValidLoggerCall { get; private set; }
+            public ImmutableArray<ExpressionSyntax> InvalidLoggerInvocations => invalidIvocationsBuilder.ToImmutableArray();
```

Comment:
I kept some of the fields for the moment. We need to settle on "KnownTypes" handling and ILanguageFacade first.

Reply:
There were few more ideas in it:
- If it's `private`, there's no reason for it to be a field.
- The current `InvalidLoggerInvocationLocations` doesn't need an explicit backing field. It can be just public auto property. Because it's just a private nested class.